### PR TITLE
[WiP][JENKINS-35459] Avoid duplicated elements and provide the full URL  on items which are using a View

### PR DESCRIPTION
**Pull request by fbelzunc**
1. Fresh Jenkins instance
2. Install the Dashboard View plugin
3. Create Folder on the instance with a job inside the folder.
4. Configure the Dashboard View to provide a left and right view showing Jenkins jobs list and Latest build and place yourself on the Dashboard View. Then, you get the following:

screen shot 2016-06-08 at 10 05 48

As you can see two elements appear (both are the same elements) and when you try to go to 1`-PetClinic-app-artifact` then you are redirected to `https://<JENKINS_URL>/job/1-PetClinic-app-artifact/` - which doesn't exist because you are inside a folder.

@reviewbybees

### CODE REVIEW

**Comment by olivergondza** [comment]
IIUC, `getSearchUrl` does not work correctly w.r.t. folders. Is that not the thing that should get fixed?

**Comment by fbelzunc** [comment]
@olivergondza I am not sure about it, because if you do getSearchUrl to return getUrl instead of getShortUrl, then any search works correctly even without the Dashboard View plugin.

**Comment by fbelzunc** [comment]
@olivergondza I will check better - to try to better understand why it happens.

**Comment by oleg-nenashev** [comment]
:bug: I also agree that getSearchUrl() should be fixed. Current approach impacts search pluggability at least.

**Comment by fbelzunc** [comment]
@oleg-nenashev WDYT about this? It fixes the issue.

File: core/src/main/java/hudson/search/SuggestedItem.java [hunk 1]
```diff
@@ -102,9 +102,6 @@ public static SuggestedItem build(SearchableModelObject searchContext, SearchIte
     }
 
     private void getUrl(StringBuilder buf) {
-        if(parent!=null) {
-            parent.getUrl(buf);
-        }
```
Comment (jglick) [inline core/src/main/java/hudson/search/SuggestedItem.java]: This smells like an incompatible change.

**Comment by jglick** [comment]
```
org.junit.ComparisonFailure: expected:</[abc/]def-ghi> but was:</[]def-ghi>
	at hudson.search.SearchTest.findAndSuggest(SearchTest.java:58)
```

At the very least needs new test coverage to reproduce the problem.

From memory and a brief inspection I doubt this is the right fix. Check `Functions.getNearestAncestorUrl` and `AbstractItem.getUrl` to start.

**Comment by fbelzunc** [comment]
@oleg-nenashev @jglick Nevermind, I know how to do the proper fix here.

I will ping you once it is done.

**Comment by fbelzunc** [comment]
@jglick @oleg-nenashev I think this is the right fix - However, I don't understand why the test fails with:

```
com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException: 404 Not Found for http://localhost:60542/jenkins/search/?q=folder%20myJob
```

When the test is running, if I launch:

```
curl -I -v http://localhost:60542/jenkins/search?q=folder%20myJob
````

Then, I am getting `302 Found`

I think this is an issue with JenkinsRule - what do you think?

**Comment by fbelzunc** [comment]
```
<
* Connection #0 to host localhost left intact
Felixs-MBP:groovy-script-useful-jenkins fbelzunc$ curl -v http://localhost:53726/jenkins/search?q=folder%20myJob
*   Trying ::1...
* connect to ::1 port 53726 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 53726 (#0)
> GET /jenkins/search?q=folder%20myJob HTTP/1.1
> Host: localhost:53726
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 302 Found
< Date: Wed, 22 Jun 2016 08:52:22 GMT
< X-Content-Type-Options: nosniff
< Stapler-Trace-001: -> evaluate(<hudson.model.Hudson@55aae7af> :hudson.model.Hudson,"/search")
< Stapler-Trace-002: -> evaluate(((StaplerProxy)<hudson.model.Hudson@55aae7af>).getTarget(),"/search")
< Stapler-Trace-003: -> evaluate(<hudson.model.Hudson@55aae7af>.getSearch(),"")
< Stapler-Trace-004: -> evaluate(<hudson.search.Search@74502066> :hudson.search.Search,"")
< Location: http://localhost:53726/jenkins/search/?q=folder%20myJob
< Content-Length: 0
< Server: Jetty(9.2.15.v20160210)
```

**Comment by fbelzunc** [comment]
Same happens with:

``
HtmlPage searchResult = wc.search(myMockFolder.getName() + " " + freeStyleProject.getName());
```

**Comment by oleg-nenashev** [comment]
Needs investigation

**Comment by oleg-nenashev** [comment]
Assigned to myself. I have zero memory of this PR btw

**Comment by oleg-nenashev** [comment]
Integrated by #2994

**Pull request closed without merging**